### PR TITLE
Title Suggestion

### PR DESCRIPTION
**Pull request by scottmacphersonmusic**
Further refinements to the automatic title suggestion feature.

### CODE REVIEW

File: lib/page_parser.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      end
+    end
+
+    private
+
+    def page
+      return @page if @page
+
+      begin
+        @page = Page.get(@url)
```
Comment (scottmacphersonmusic) [inline lib/page_parser.rb:32]: @bion I've been trying to stub `PageParser::Page.get` (I've also tried it as `PageParser::HtmlParserIncluded.get`) to respond with a local html file but haven't had any luck.  Short of that I'm having trouble thinking of any valuable way to test this module.

File: app/assets/javascripts/auto_title.js [hunk 2]
```diff
@@ -7,7 +7,7 @@ function path(name) {
 }
 
 function retrieveTitleSuggestion() {
-  var link = this.value;
+  var link = encodeURI(this.value);
```
Comment (scottmacphersonmusic) [inline app/assets/javascripts/auto_title.js:10]: @bion adding `encodeURI()` is my attempt at protecting against XSS in the form input.  

You also mentioned being careful with what gets returned when the `<title>` element is parsed out, but as I recall when we last talked about XSS, Rails automatically escapes any dangerous characters with the `sanitize` method.

File: lib/page_parser.rb (hunk 1 above)
Comment (bion) [inline lib/page_parser.rb:32]: the test where you're doing this:

``` ruby
    PageParser::LinkInfo.any_instance
                        .stubs(:page)
                        .returns(Nokogiri::HTML(markup))
```

is passing for me.

File: app/assets/javascripts/auto_title.js (hunk 2 above)
Comment (bion) [inline app/assets/javascripts/auto_title.js:10]: Encoding the URL on the client side doesn't change the security situation b/c the user can do whatever they want with your javascript. Someone could very easily open the development console of their browser and remove the call to `encodeURI` if they so choose. Protection against XSS needs to happen on the server, which, like you mentioned, is being sure to sanitize all strings that you send back to the client for display.

**Pull request merged**